File: backend/memory/lessons.py
from typing import Dict, Any
from .rag import RAGSystem
from ..database.adapter import DatabaseAdapter
from ..core.ai_client import AIClient

class LessonManager:
    def __init__(self, db: DatabaseAdapter, ai_client: AIClient):
        self.db = db
        self.ai_client = ai_client

    def process_closed_trade(self, trade_id: str):
        # 1. Get trade info
        with self.db.engine.connect() as conn:
            result = conn.execute(self.db.trades.select().where(self.db.trades.c.id == trade_id))
            trade = result.fetchone()
            if not trade:
                return
            trade = dict(trade._asdict())

        # 2. Analyze outcome (Already has PnL in DB from execute_sell)
        outcome = {
            "pnl": trade['pnl'],
            "pnl_porcentaje": trade['pnl_porcentaje'],
            "contexto_mercado": f"Precio salida: {trade['precio_salida']} el {trade['fecha_salida']}"
        }

        # 3. Generate lesson with AI
        try:
            lesson_data = self.ai_client.generate_lesson(trade, outcome)
            
            # 4. Save to DB
            lesson_to_save = {
                "id": str(uuid.uuid4()),
                "trade_id": trade_id,
                "ticker": trade['ticker'],
                "fecha": datetime.now().isoformat(),
                **lesson_data
            }
            # Convert tags list to string if necessary
            if isinstance(lesson_to_save.get('tags'), list):
                lesson_to_save['tags'] = ",".join(lesson_to_save['tags'])
                
            self.db.insert_lesson(lesson_to_save)
            
            # Mark trade as having a lesson generated
            self.db.update_trade(trade_id, {"leccion_generada": 1})
        except Exception as e:
            print(f"Error generando lección para {trade_id}: {e}")

File: frontend/lib/lessons.ts
/**
 * Memoria de largo plazo: port de backend/memory/{lessons,rag}.py
 */

import { v4 as uuidv4 } from "uuid";
import { getServerSupabase } from "./supabase/server";
import { generateLesson } from "./groq";

export async function getLessonsByTicker(
  ticker?: string,
  limit = 5
): Promise<any[]> {
  const db = getServerSupabase();
  let q = db
    .from("lecciones")
    .select("*")
    .order("fecha", { ascending: false })
    .limit(limit);
  if (ticker) q = q.eq("ticker", ticker);
  const { data, error } = await q;
  if (error) throw error;
  return data ?? [];
}

export async function getRelevantMemory(ticker: string): Promise<string> {
  const lessons = await getLessonsByTicker(ticker, 5);
  if (!lessons.length) return "No hay antecedentes específicos para este ticker en la memoria.";
  return lessons
    .map(
      (l) =>
        `[${l.fecha}] ${l.titulo ?? "Lección"}: ${l.leccion ?? ""} (Patrón: ${l.patron ?? "N/A"})`
    )
    .join("\n\n");
}

export async function processClosedTrade(tradeId: string): Promise<void> {
  const db = getServerSupabase();
  const { data: trade, error } = await db
    .from("trades")
    .select("*")
    .eq("id", tradeId)
    .single();
  if (error || !trade) return;
  if (trade.leccion_generada) return;

  try {
    const outcome = {
      pnl: trade.pnl ?? 0,
      pnl_porcentaje: trade.pnl_porcentaje ?? 0,
      contexto_mercado: `Precio salida: ${trade.precio_salida} el ${trade.fecha_salida}`,
    };
    const lesson = await generateLesson(trade, outcome);

    await db.from("lecciones").insert({
      id: uuidv4(),
      trade_id: tradeId,
      ticker: trade.ticker,
      fecha: new Date().toISOString(),
      tipo: lesson.tipo,
      titulo: lesson.titulo,
      leccion: lesson.leccion,
      tags: Array.isArray(lesson.tags) ? lesson.tags : [],
      patron: lesson.patron,
      aplicar_cuando: lesson.aplicar_cuando,
      estrellas: lesson.estrellas,
    });

    await db.from("trades").update({ leccion_generada: true }).eq("id", tradeId);
  } catch (e) {
    console.error("Error generando lección para", tradeId, e);
  }
}
